fast64_internal/f3d/f3d_material.py

```diff
@@ -1722,39 +1722,47 @@ def update_tex_values_manual(material: bpy.types.Material, context, prop_path=No
     set_texture_settings_node(material)
 
 
-def getMaterialScrollDimensions(material):
-    useDict = all_combiner_uses(material)
-
-    if (
-        useDict["Texture 0"]
-        and material.tex0.tex is not None
-        and useDict["Texture 1"]
-        and material.tex1.tex is not None
-        and material.tex0.tex.size[0] > 0
-        and material.tex0.tex.size[1] > 0
-        and material.tex1.tex.size[0] > 0
-        and material.tex1.tex.size[1] > 0
-    ):
-        if material.uv_basis == "TEXEL0":
-            return material.tex0.tex.size
-        else:
-            return material.tex1.tex.size
-    elif (
-        useDict["Texture 1"]
-        and material.tex1.tex is not None
-        and material.tex1.tex.size[0] > 0
-        and material.tex1.tex.size[1] > 0
-    ):
-        return material.tex1.tex.size
-    elif (
-        useDict["Texture 0"]
-        and material.tex0.tex is not None
-        and material.tex0.tex.size[0] > 0
-        and material.tex0.tex.size[1] > 0
-    ):
-        return material.tex0.tex.size
+def shift_num(num: int, amt: int):
+    if amt < 0:
+        return num >> -amt
+    return num << amt
+
+
+def shift_dimensions(tex_prop: "TextureProperty", dimensions: tuple[int, int]):
+    shifted = (shift_num(dimensions[0], tex_prop.S.shift), shift_num(dimensions[1], tex_prop.T.shift))
```

Comment:
About this

The texture is mapped in ST space from `0,0` to `dimensions[0],dimensions[1]`, right

When the scrolling sets a ST position `s,t`, that position becomes `s<<shiftS,t<<shiftT` (assume `<<-x ~ >>x` for simplicity), in the RDP, right

So in the end the scrolling code should ensure `s<<shiftS,t<<shiftT` ranges from `0,0` to `dimensions[0],dimensions[1]` (the ST space range the texture is mapped to), right

That would mean the max for `s,t` should be
```suggestion
    shifted = (shift_num(dimensions[0], -tex_prop.S.shift), shift_num(dimensions[1], -tex_prop.T.shift))
```
instead?

(I'm really not sure. I'm more asking for my own knowledge)

Reply:
@Dragorn421 this is specifically for UV scrolling, you are correct with tile scrolling though, although you can set S/T low to numbers higher than the dimensions without any issue. also S/T low is pre-shift/scaling, so if you have a 32/32 texture and set S low to 32, that is always going to be the full width of the texture

Reply:
Uv scrolling and tile scrolling isn't the same thing?

Reply:
@Dragorn421 nope, tile can be per texture and thats how you get pretty water

Reply:
The PR is correct on this, I was wrong on shiftS being used as a left shift, it's a right shift